src/vm/checker/typecheck/natives/assets.rs

```diff
@@ -0,0 +1,166 @@
+        Ok(AtomTypeIdentifier::OptionalType(
+            Box::new(AtomTypeIdentifier::PrincipalType.into())).into())
+    }
+}
+
+pub fn check_special_get_balance(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(2, args)?;
+
+    let asset_name = args[0].match_atom()
+        .ok_or(CheckErrors::BadAssetName)?;
+    checker.type_map.set_type(&args[0], no_type())?;
+
+    if !checker.contract_context.token_exists(asset_name) {
+        return Err(CheckErrors::NoSuchAsset(asset_name.clone()).into());
+    }
+
+    let supplied_owner_type = checker.type_check(&args[1], context)?;
+    let expected_owner_type: TypeSignature = AtomTypeIdentifier::PrincipalType.into();
+
+    if !expected_owner_type.admits_type(&supplied_owner_type) {
+        Err(CheckErrors::TypeError(expected_owner_type, supplied_owner_type).into())
+    } else {
+        Ok(AtomTypeIdentifier::IntType.into())
+    }
+}
+
+pub fn check_special_mint_asset(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(3, args)?;
+
+    let asset_name = args[0].match_atom()
+        .ok_or(CheckErrors::BadAssetName)?;
+    checker.type_map.set_type(&args[0], no_type())?;
+
+    let supplied_asset_type = checker.type_check(&args[1], context)?;
+    let supplied_owner_type = checker.type_check(&args[2], context)?;
+
+    let expected_owner_type: TypeSignature = AtomTypeIdentifier::PrincipalType.into();
+    let expected_asset_type = checker.contract_context.get_asset_type(asset_name)
+        .ok_or(CheckErrors::NoSuchAsset(asset_name.clone()))?;
+
+    if !expected_asset_type.admits_type(&supplied_asset_type) {
+        return Err(CheckErrors::TypeError(expected_asset_type.clone(), supplied_asset_type).into())
+    }
+
+    if !expected_owner_type.admits_type(&supplied_owner_type) {
+        return Err(CheckErrors::TypeError(expected_owner_type, supplied_owner_type).into())
+    }
+
+    Ok(AtomTypeIdentifier::ResponseType(
+        Box::new((AtomTypeIdentifier::BoolType.into(),
+                  AtomTypeIdentifier::IntType.into()))).into())
+}
+
+pub fn check_special_mint_token(checker: &mut TypeChecker, args: &[SymbolicExpression], context: &TypingContext) -> TypeResult {
+    check_argument_count(3, args)?;
+
+    let asset_name = args[0].match_atom()
+        .ok_or(CheckErrors::BadAssetName)?;
+    checker.type_map.set_type(&args[0], no_type())?;
+
+    let supplied_amount = checker.type_check(&args[1], context)?;
+    let supplied_owner_type = checker.type_check(&args[2], context)?;
+
+    let expected_amount: TypeSignature = AtomTypeIdentifier::IntType.into();
+    let expected_owner_type: TypeSignature = AtomTypeIdentifier::PrincipalType.into();
+
+    if !checker.contract_context.token_exists(asset_name) {
+        return Err(CheckErrors::NoSuchAsset(asset_name.clone()).into());
+    }
+
+    if !expected_amount.admits_type(&supplied_amount) {
```

Comment:
Not sure if this belongs in the type check, but do we want to make sure that the integer value supplied is positive?  Same goes for `check_special_transfer_token()`

Reply:
The integer being positive can't be type checked at the moment -- we only have signed integers.